src/System.IO.Compression/src/System/IO/Compression/DeflaterZLib.cs

```diff
@@ -0,0 +1,280 @@
+            int memLevel;
+            ZLibNative.CompressionStrategy strategy;
+
+            switch (compressionLevel)
+            {
+                // Note that ZLib currently exactly correspond to the optimal values.
+                // However, we have determined the optimal values by intependent measurements across
+                // a range of all possible ZLib parameters and over a set of different data.
+                // We stress that by using explicitly the values obtained by the measurements rather than
+                // ZLib defaults even if they happened to be the same.
+                // For ZLib 1.2.3 we have (copied from ZLibNative.cs):
+                // ZLibNative.CompressionLevel.DefaultCompression = 6;
+                // ZLibNative.Deflate_DefaultWindowBits = -15;
+                // ZLibNative.Deflate_DefaultMemLevel = 8;
+
+
+                case CompressionLevel.Optimal:
+                    zlibCompressionLevel = (ZLibNative.CompressionLevel)6;
+                    windowBits = -15;
+                    memLevel = 8;
+                    strategy = ZLibNative.CompressionStrategy.DefaultStrategy;
+                    break;
+
+                case CompressionLevel.Fastest:
+                    zlibCompressionLevel = (ZLibNative.CompressionLevel)1;
+                    windowBits = -15;
+                    memLevel = 8;
+                    strategy = ZLibNative.CompressionStrategy.DefaultStrategy;
+                    break;
+
+                case CompressionLevel.NoCompression:
+                    zlibCompressionLevel = (ZLibNative.CompressionLevel)0;
+                    windowBits = -15;
+                    memLevel = 7;
+                    strategy = ZLibNative.CompressionStrategy.DefaultStrategy;
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException("compressionLevel");
+            }
+
+            _isDisposed = false;
+            DeflateInit(zlibCompressionLevel, windowBits, memLevel, strategy);
+        }
+
+        void IDisposable.Dispose()
+        {
+            Dispose(true);
+        }
+
+        [SecuritySafeCritical]
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing && !_isDisposed)
+            {
+                if (_inputBufferHandle.HasValue)
+                    DeallocateInputBufferHandle();
+
+                _zlibStream.Dispose();
+                _isDisposed = true;
+            }
+        }
+
+        private bool NeedsInput()
+        {
+            // Convenience method to call NeedsInput privately without a cast.
+            return ((IDeflater)this).NeedsInput();
+        }
+
+        bool IDeflater.NeedsInput()
+        {
+            return 0 == _zlibStream.AvailIn;
+        }
+
+        void IDeflater.SetInput(byte[] inputBuffer, int startIndex, int count)
+        {
+            Contract.Assert(NeedsInput(), "We have something left in previous input!");
```

Comment:
We haven't been, though we also haven't been adding more.  We primarily just decided to use Debug.Assert consistently across the codebase instead of Contract.Assert.  Our other usage of contracts is currently in a limbo state; we don't run the contract checker, so they've gotten stale, and we've not decided how we want to proceed with them.